File: src/go/luhn/luhn.go
package luhn

func Check(n int64) bool {
	var digits []int
	for n > 0 {
		digits = append([]int{int(n % 10)}, digits...)
		n /= 10
	}

	var checksum int
	for i, d := range digits {
		if i%2 == len(digits)%2 {
			d *= 2
			if d > 9 {
				d -= 9
			}
		}
		checksum += d
	}

	return checksum%10 == 0
}

File: src/kotlin/luhn/Luhn.kt
package luhn

fun check(n: Long): Boolean {
    val digits = n.toString().map { it.toString().toInt() }.reversed()
    var checksum = 0
    var even = false

    for (d in digits) {
        val digit = d * if (even) 2 else 1
        checksum += if (digit > 9) digit - 9 else digit
        even = !even
    }

    return checksum % 10 == 0
}
